server/handlers/default.go

```diff
@@ -143,70 +143,95 @@ func DeleteHandler(ctx context.Context, w http.ResponseWriter, r *http.Request)
+}
+
+type serverKeyInfo struct {
+	gun     string
+	role    string
+	store   storage.MetaStore
+	crypto  signed.CryptoService
+	keyAlgo string
 }
 
-func getKeyHandler(ctx context.Context, w http.ResponseWriter, r *http.Request, vars map[string]string) error {
+func parseKeyParams(ctx context.Context, vars map[string]string) (*serverKeyInfo, error) {
 	gun, ok := vars["imageName"]
 	if !ok || gun == "" {
-		return errors.ErrUnknown.WithDetail("no gun")
+		return nil, errors.ErrUnknown.WithDetail("no gun")
 	}
 	role, ok := vars["tufRole"]
 	if !ok || role == "" {
-		return errors.ErrUnknown.WithDetail("no role")
+		return nil, errors.ErrUnknown.WithDetail("no role")
+	}
+	if role != data.CanonicalTimestampRole && role != data.CanonicalSnapshotRole {
+		return nil, errors.ErrInvalidRole.WithDetail(role)
 	}
-
-	logger := ctxu.GetLoggerWithField(ctx, gun, "gun")
 
 	s := ctx.Value("metaStore")
 	store, ok := s.(storage.MetaStore)
 	if !ok || store == nil {
-		logger.Error("500 GET storage not configured")
-		return errors.ErrNoStorage.WithDetail(nil)
+		return nil, errors.ErrNoStorage.WithDetail("metadata store not configured")
 	}
+
 	c := ctx.Value("cryptoService")
 	crypto, ok := c.(signed.CryptoService)
 	if !ok || crypto == nil {
-		logger.Error("500 GET crypto service not configured")
-		return errors.ErrNoCryptoService.WithDetail(nil)
+		return nil, errors.ErrNoCryptoService.WithDetail("crypto service not configured")
 	}
+
 	algo := ctx.Value("keyAlgorithm")
 	keyAlgo, ok := algo.(string)
-	if !ok || keyAlgo == "" {
-		logger.Error("500 GET key algorithm not configured")
-		return errors.ErrNoKeyAlgorithm.WithDetail(nil)
+	if !ok || keyAlgo != data.ECDSAKey && keyAlgo != data.RSAKey && keyAlgo != data.ED25519Key {
+		return nil, errors.ErrNoKeyAlgorithm.WithDetail("key algorithm not configured")
+	}
+
+	return &serverKeyInfo{
+		gun:     gun,
+		role:    role,
+		store:   store,
+		crypto:  crypto,
+		keyAlgo: keyAlgo,
+	}, nil
+}
+
+func rotateKeyHandler(ctx context.Context, w io.Writer, vars map[string]string) error {
+	return getKeyHandler(ctx, w, vars)
+}
+
+func getKeyHandler(ctx context.Context, w io.Writer, vars map[string]string) error {
+	s, err := parseKeyParams(ctx, vars)
+	if err != nil {
+		return err
 	}
-	keyAlgorithm := keyAlgo
+	var key data.PublicKey
 
-	var (
-		key data.PublicKey
-		err error
-	)
-	switch role {
+	switch s.role { // parseKeyParams ensures it's only timestamp or snapshot
 	case data.CanonicalTimestampRole:
-		key, err = timestamp.GetOrCreateTimestampKey(gun, store, crypto, keyAlgorithm)
+		key, err = timestamp.GetOrCreateTimestampKey(s.gun, s.store, s.crypto, s.keyAlgo)
 	case data.CanonicalSnapshotRole:
-		key, err = snapshot.GetOrCreateSnapshotKey(gun, store, crypto, keyAlgorithm)
-	default:
-		logger.Errorf("400 GET %s key: %v", role, err)
-		return errors.ErrInvalidRole.WithDetail(role)
+		key, err = snapshot.GetOrCreateSnapshotKey(s.gun, s.store, s.crypto, s.keyAlgo)
```

Comment:
I don't mind moving it back, but I was going to have a `rotateKeyHandler`, which would need to perform the same check, so if we want it in a single location, should it be in `parseKeys`?  `https://github.com/docker/notary/pull/534/files#diff-b60e09f129d59867e23c3c6e1e94319cR159` was eventually supposed to be how this would look.

I can also just move it back to `getKeyHandler` for now, and when/if that change lands, move it to `parseKeyParams`